File: leetcode_JS/arr_MoveZeroes.js
/*
https://leetcode.com/problems/move-zeroes/description/
Given array, move all 0s to end while maintaining relative order of non-zero elements
Example, given nums = [0, 1, 0, 3, 12], ans should be [1, 3, 12, 0, 0]

Do this in-place without making a copy of the array
Minimize the total number of operations
*/

// Start from the back so we wouldnt have to worry about index at the start 
moveZeroes = (nums) => {  
    for (let i=nums.length-1 ; i>=0 ; i--) {
        if (nums[i] === 0) {
            nums.splice(i, 1);
            nums.push(0);
        }
    }
};

nums = [1, 3, 12, 0, 0];
moveZeroes(nums);
console.log(nums); // [ 1, 3, 12, 0, 0 ]

File: leetcode_Python/arr_MoveZeroes.py
'''
https://leetcode.com/problems/move-zeroes/description/
Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.
'''

def move_short(nums):
    zeroIndex = 0
    for i, num in enumerate(nums):
        if num != 0:
            nums[zeroIndex], nums[i] = nums[i], nums[zeroIndex]
            zeroIndex += 1

arr = [0, 0, 1]
move_short(arr)
print(arr) # [1, 0, 0]

def moveZeroes(nums):
    """
    :type nums: List[int]
    :rtype: void Do not return anything, modify nums in-place instead.
    """
    start, end, count = 0, len(nums), 0
    while start < end:
        if nums[start] == 0:
            nums[:] = nums[:start] + nums[start+1:] + [nums[start]]
            count += 1 
            start -= 1
        if start + count == end: 
            break
        start += 1

arr = [0, 1, 0, 3, 12]
moveZeroes(arr)
print(arr) # [1, 3, 12, 0, 0]
